reflect.go

```diff
@@ -114,8 +114,14 @@ type ConcreteType struct {
 	Byte byte
 }
 
-// Must use this to register an interface to properly decode the
-// underlying concrete type.
+// This function should be used to register the receiving interface that will
+// be used to decoded an underlying concrete type. If embedding the interface
+// in a struct, the interface MUST be the only field and it MUST be an exported field.
+// For example:
+//	RegisterInterface(struct{ Animal }{}, ConcreteType{&foo, 0x01)
+// OR
+//	type exportedField struct { Animal animal }
+//	RegisterInterface(exportedField{}, ConcreteType{&foo, 0x01)
```

Comment:
Embedding the interface in a struct is required, not optional.
The purpose of `struct { MyInterface }` is simply to make it as easy as possible to pass in an interface type to the function.  The alternative is to pass a nil pointer to the interface, such as (*MyInterface)(nil).  See https://stackoverflow.com/questions/7132848/how-to-get-the-reflect-type-of-an-interface , see how there isn't a direct way to do this.

`type exportedField struct { Animal animal }` is thus an antipattern... there's no point to naming this type since it won't be re-used, and all you've done is turn a 1 liner into 2 lines.

Reply:
@jaekwon in deed, the proper usage pattern is well encapsulated in the first RegisterInterface example. Would you prefer we scrape the second/OR usage?

Reply:
if that's the antipattern, my answer would be yes). we should probably modify tests too to reflect this